### PR TITLE
[Bugfix][V1] Fix allowed_token_ids for v1 Sampler

### PR DESCRIPTION
**Pull request by houseroad**
Revert the one and zero, so masked_fill_ is applied appropriately.

Tested with 'test_allowed_token_ids' in https://github.com/vllm-project/vllm/pull/14159/

Additional test:

```
    batched_output = xx.generate(
    ¦   [PROMPT, PROMPT],
    ¦   [
    ¦   ¦   SamplingParams(allowed_token_ids=allowed_token_ids),
    ¦   ¦   SamplingParams(),
    ¦   ]
    )
    assert batched_output[1].outputs[0].token_ids[-1] != TOKEN_ID
```

### CODE REVIEW

File: vllm/v1/engine/processor.py [hunk 1]
```diff
@@ -92,10 +92,13 @@ def _validate_allowed_token_ids(
             return
         if params.allowed_token_ids is None:
             return
-        if not all(0 <= tid < self.model_config.vocab_size
+        if params.allowed_token_ids is not None and len(
+                params.allowed_token_ids) == 0:
```
Comment (ywang96) [inline vllm/v1/engine/processor.py:96]: This can be just `if not params.allowed_token_ids` since if `params.allowed_token_ids is None` it would have already returned above.

File: vllm/v1/worker/gpu_input_batch.py [hunk 2]
```diff
@@ -300,17 +300,17 @@ def add_request(
...
+                                                         dtype=torch.bool,
+                                                         device=self.device)
+                self.allowed_token_ids_mask_cpu_tensor = torch.ones(
                     self.max_num_reqs,
                     self.vocab_size,
                     dtype=torch.bool,
                     device="cpu")
             self.allowed_token_ids_mask_cpu_tensor[req_index][
-                sampling_params.allowed_token_ids] = True
+                sampling_params.allowed_token_ids] = False
```
Comment (ywang96) [inline vllm/v1/worker/gpu_input_batch.py:313]: I think the updated logic here is a bit counter-intuitive to readers from a glance - can we add a `NOTE` here that all token ids with `True` mask will be filled with `float(-inf)` during sampling?

**Review by ywang96 — APPROVED** [review]
LGTM! Thanks for fixing this and I left two minor comments!

File: vllm/v1/engine/processor.py [hunk 3]
```diff
@@ -92,10 +92,13 @@ def _validate_allowed_token_ids(
...
+            raise ValueError("allowed_token_ids is not None and empty!")
+        if not all(0 <= tid < self.model_config.get_vocab_size()
```
Comment (njhill) [inline vllm/v1/engine/processor.py:98]:
```suggestion
        if not params.allowed_token_ids:
            raise ValueError("allowed_token_ids cannot be empty")
        vocab_size = self.model_config.get_vocab_size()
        if not all(0 <= tid < vocab_size
```

**Review by njhill — COMMENTED** [review]
Thanks @houseroad, looks good.

File: vllm/v1/engine/processor.py (hunk 3 above)
Comment (houseroad) [reply vllm/v1/engine/processor.py]: Make sense, addressed the comments

**Review by ywang96 — CHANGES_REQUESTED** [review]
Please check the discussion on slack regarding `self.allowed_token_ids_mask` - we need to take into account the situation where requests have different sampling params, which is fairly common in the online setting.

**Comment by njhill** [comment]
@liangfu yes thanks to @robertgshaw2-redhat for pointing out we only want to invert the mask for the applicable rows. So the tensors should still be initialized to zeros / reset to zeros when the request is removed, and when adding the requests, just fill the row to 1's before setting the allowed tokens to zero.

File: vllm/v1/worker/gpu_input_batch.py [hunk 4]
```diff
@@ -359,7 +362,7 @@ def remove_request(self, req_id: str) -> Optional[int]:
         self.logit_bias[req_index] = None
         self.has_allowed_token_ids.discard(req_id)
         if self.allowed_token_ids_mask_cpu_tensor is not None:
-            self.allowed_token_ids_mask_cpu_tensor[req_index].fill_(False)
+            self.allowed_token_ids_mask_cpu_tensor[req_index].fill_(True)
```
[the remark this answers is not in the document]
Comment (houseroad) [reply vllm/v1/worker/gpu_input_batch.py]: True, added more comments to help understand

**Review by njhill — APPROVED** [review]
Thanks @houseroad, LGTM now

**Review by ywang96 — APPROVED** [review]
Thanks for the fix!
